Flesh out glossary

s-ben opened the pull request:
Fleshes out 1st draft of glossary.

I think we have solid definitions for all terms listed now. There are a handful of definitions I've flagged for review with `(SETH: comment...)`.

Propose we make the **bold** terms into subheadings. This should look nice and also allow for relative links between definitions.

Commit:
Fills in some definitions, tweaks wording

Commit:
Finishes fleshing out definitions, tweaks for flow

RichardRed0x commented on README.md:
```diff
@@ -2,59 +2,86 @@ dcrwords
 
-Wallet: 
+**Wallet:**
+A file that stores private keys. Wallets typically contain a software client which allows its owner to view and create transactions on the  blockchain. Decred's native wallet, [Decredition](https://docs.decred.org/getting-started/user-guides/decrediton-setup/) is available in GUI and Command Line Interface (CLI) form (for advanced users). Decred also has a [web wallet](https://docs.decred.org/get ...
```
Decrediton is the GUI, CLI is not Decrediton

s-ben replied:
Updated with the following text:

A file that stores private keys. Wallets typically contain a software client which allows its owner to view and create transactions on the  blockchain. Decred's native wallet daemon, can be accesseed via [Decredition](https://docs.decred.org/getting-started/user-guides/decrediton-setup/), which provides a GUI interface, as well as the Command Line Interface (CLI) form (for advanced users). Decred also has a [web wallet](https://docs.decred.org/getting-started/user-guides/web/) and is supported by numerous [3rd party wallets](https://decred.org/downloads/).

RichardRed0x replied:
"contain a software client" - odd wording.
The web wallet is being phased out I believe so probably shouldn't feature here.

s-ben replied:
Updated wording to:

Wallet: A file that stores private keys. Wallets allow its owner to view and create transactions on the blockchain. Decred's native wallet daemon can be accesseed via Decredition, which provides a GUI interface, as well as the Command Line Interface (CLI) wallet (for advanced users). Decred is also supported by numerous 3rd party wallets.

RichardRed0x replied:
"its owner" -> "their owner"
Would suggest simplifying the final sentence to: "The Decred team maintains a GUI wallet (Decrediton) and Command Line Interface (CLI) wallet tools - for advanced users. Decred is also supported by numerous 3rd party wallets."

s-ben replied:
Going with your suggestions. I think this one is good to go:

"A file that stores private keys. Wallets allow their owner to view and create transactions on the blockchain. The Decred team maintains a GUI wallet (Decrediton) and Command Line Interface (CLI) wallet tools for advanced users. Decred is also supported by numerous 3rd party wallets."

RichardRed0x commented on README.md:
```diff
@@ -2,59 +2,86 @@ dcrwords
 
-**Account:** A single Decred wallet can be used to operate multiple independent accounts (i.e. DCR transfers between accounts are on-chain transactions).
+**Account:** 
+A single Decred wallet can be used to operate multiple independent accounts (i.e. DCR transfers between accounts are on-chain transactions).(SETH: Have not seen the concept of an 'account' before, outside of Politeia. Don't see it in the dcrdocs for wallets...Is a wallet/account just software that controls multiple addresses? Related to HD wallets?)
```
Have a look in Decrediton, try it out.

s-ben replied:
Just going with copy in Decredition for now. It's accurate at least. If someone wants to improve they can:

Accounts allow you to keep separate records of your DCR funds. Transferring DCR across accounts will create a transaction on the blockchain.

RichardRed0x commented on README.md:
```diff
@@ -2,59 +2,86 @@ dcrwords
 
-Proof-of-Stake voting:
+**Proof-of-Stake (PoS) voting:**
+The mechanism by which ticket holders vote to validate blocks confirmed by PoW miners (thus providing a check on PoW miners), earn staking rewards, and vote on consensus rule changes and Politeia proposals. To vote, a ticket holder's wallet must be online when called to vote. Ticket holders may vote using their own personal wallet or utilize Voting Service Providers (VSP) (stakepools) to vote for ...
```
We are trying not to use "validate" in relation to PoS block voting because it is not strictly accurate. Preferred term is "approve".
I'm not sure the VSP stuff belongs here.

s-ben replied:
Used "approved", removed all but basic definition:

The mechanism by which ticket holders vote to approve blocks confirmed by PoW miners (thus providing a check on PoW miners), earn staking rewards, and vote on consensus rule changes and Politeia proposals.

RichardRed0x commented on README.md:
```diff
@@ -2,59 +2,86 @@ dcrwords
+The mechanism by which ticket holders vote to validate blocks confirmed by PoW miners (thus providing a check on PoW miners), earn staking rewards, and vote on consensus rule changes and Politeia proposals. To vote, a ticket holder's wallet must be online when called to vote. Ticket holders may vote using their own personal wallet or utilize Voting Service Providers (VSP) (stakepools) to vote for ...
 
 **[Hybrid PoW/PoS:](https://docs.decred.org/research/hybrid-design/)** 
+In Decred, a [hybrid PoW/PoS system](https://docs.decred.org/research/hybrid-design/) is used, whereby blocks mined by PoW miners must be validated by ticket holders, which are pseudorandomly selected from the ticket pool. This provides a check on PoW miners and increases the overall cost of attacking the network. When a block is mined, 60% of the block reward goes to the PoW miner, 30% goes to t ...
```
validated -> approved

s-ben replied:
Replaced validated with approved:

In Decred, a [hybrid PoW/PoS system](https://docs.decred.org/research/hybrid-design/) is used, whereby blocks mined by PoW miners must be approved by ticket holders, which are pseudorandomly selected from the ticket pool. This provides a check on PoW miners and increases the overall cost of attacking the network. When a block is approved, 60% of the block reward goes to the PoW miner, 30% goes to the holders of  tickets called to vote (approve blocks), and 10% goes to the Decred Treasury to fund the project's development.

RichardRed0x commented on README.md:
```diff
@@ -2,59 +2,86 @@ dcrwords
 
-Block reward:
+**Block reward:**
+A reward (in DCR) given to PoW miners for successfully validating a new block of transactions. As in Bitcoin, the block reward [decays exponentially](https://docs.decred.org/advanced/inflation/) with block height (length of blockchain). However, Decred’s algorithm interpolates this decay over time so as not to produce market shocks, as typically seen when Bitcoin's difficulty halves. The block re ...
```
Block reward is shared by PoW/PoS/Treasury

s-ben replied:
Updated to:

A reward (in DCR) split between PoW miners, ticket holders, and the Treasury for successfully approving a block of transactions. As in Bitcoin, the block reward [decays exponentially](https://docs.decred.org/advanced/inflation/) with block height (length of blockchain). However, Decred’s algorithm interpolates this decay over time so as not to produce market shocks, as typically seen when Bitcoin's difficulty halves. The block reward started at 31.19582664 and it adjusts every 6,144 blocks (approximately 21.33 days) by reducing by a factor of 100/1011.

RichardRed0x commented on README.md:
```diff
@@ -2,59 +2,86 @@ dcrwords
 
-Stake version interval: 
+**Stake Version Interval (SVI):** 
+An interval of blocks 2016 blocks (~1 week) which is used to determine if a vote on concensus rules can begin. Before a vote on consensus rule changes can begin, 75% of tickets that vote during a Stake Version Interval (SVI) must be using software that contains the latent software change being proposed. 
```
There's also a mining node requirement, 95% iirc

s-ben replied:
I was going to include the mining node requirement, but didn't because that is independent of the SVI. I believe the interval for PoW miners to upgrade is 1,000 blocks. I do think it's relevant here, but it's a tradeoff between keeping the definition concise and "pure", and not misleading the reader by implying that rule changes can begin if 75% of tickets vote for it (without consideration by PoW miners). I tried to be careful in wording not to imply that, but can see how it might be implied anyway. Open to either course.

RichardRed0x commented on README.md:
```diff
@@ -64,81 +91,114 @@ Rule change interval:
+Transactions waiting to be mined.
 
 **Block header:** 
+A 120 byte (when serialized) header used to identify a particular block, which is hashed repeatedly to create proof of work for mining rewards. (SETH: review for accuracy)
```
"create proof of work for mining rewards" seems confusing

s-ben replied:
I agree. Got this wording from the Bitcoin.org docs I believe. Will not attempt a rewrite at this moment, as I'm not that sure how this works.

s-ben replied:
The relationship to PoW seems unnecessary confusing here. Using the simpler wording below, which I have high confidence is accurate. 

"Metadata at the beginning of a block that defines information about the block. The hash of a block header is the main way of identifying a block."

RichardRed0x commented on README.md:
```diff
@@ -64,81 +91,114 @@ Rule change interval:
 
-Proof-of-Work mining:
+**Proof-of-Work mining:**
+The process by which Proof-of-Work (PoW) miners validate blocks of transactions. In this process, miners compute hashes until the block header is below the target difficulty. (SETH: review for accuracy. Expand?)
```
Pretty sure this is not accurate, in that I don't think it's the "block header" which must be below the target difficulty. I could be wrong, not an expert on PoW by any means.

s-ben replied:
Agreed. I spent some time searching for other definitions out there, and understanding PoW for Bitcoin (basically the same). Believe this wording came from another definition, and jelled enough with my understanding to reuse it. But this should be reviewed for accuracy by someone that understands this better, and perhaps we can come up with better wording.

RichardRed0x replied:
I would say we have a difference in approach here, because I try quite hard to avoid making PRs against a repo like dcrdocs that contain errors or things I'm not confident about. It takes more of my time, but saves the time of reviewers if they're reviewing something with less errors. That's not to say my PRs are error-free, far from it, but they never include things that I'm pretty sure are wrong.

s-ben replied:
I generally agree. However, sometimes I will put stuff out for review that I know is wrong. Partly because a wrong guess sometimes is an efficient way of expressing what you're looking for. The "scaffolding" of a correct answer, which does some legwork for the accurate answer. Also, sometimes it's the only way to get a developer's attention. They can ignore a request for review (overloaded as they are), but an inaccurate definition cannot stand. It's like waving a red cape in front of a bull. They lose control. Before you know it you have a wall of text and are hopefully still in one piece. Try and avoid that as much as possible though. I think through some additional research I've been able to get 2/3 of the definitions over the line (confident enough to put in PR for review). PoW is too complex. I say let's leave it out for now.

RichardRed0x commented on README.md:
```diff
@@ -64,81 +91,114 @@ Rule change interval:
 
-Miners:
+**Miners:**
+Miners help secure the network via Proof-of-Work (PoW) mining. Miners recieve block rewards for successfully validating blocks. 
```
block rewards [and transaction fees]

s-ben replied:
Changed to:

Miners help secure the network via Proof-of-Work (PoW) mining. Miners recieve block rewards and transaction fees for successfully validating blocks.

RichardRed0x reviewed (changes requested):
Thanks, I requested some changes. 

Also, please remove the notes from the body of the file as I don't think I can selectively omit them when merging the PR.

Commit:
Incorporates Richard-Red's requested changes

Makes most changes suggested by Richard-Red. There are a handful of suggested edits that I couldn't address. Have just left comments on those on the PR.

Commit:
Updates wallet definition

RichardRed0x commented:
See my comment on the PoW mining description for why I'm hesitating to accept this PR and make a PR against dcrdocs to add the glossary page - in short, I'd rather not make a PR that includes things I'm pretty sure are inaccurate. Here are some options for how to proceed:

1. work on the few definitions which we think are inaccurate until we think they're correct
2. remove the definitions we think are probably inaccurate, leave them blank for someone else to fill
3. I could merge this PR (who cares about dcrwords, right?) and you can open a PR to merge into dcrdocs if you think that hoping reviewers catch and fix the errors is the best way forward. 

Option 2 seems solid to me, Option 1 works too if you want to take it on now, I could take it on either but not for a day or two.

Commit:
Updates wallet, block header. Removes PoW

s-ben commented:
I think I'm suggesting all three options? I was able to research further and get a better definition for 'block header'. It's high-level, but I'm confident it's correct and useful. Your rewording for 'wallet' is good, and I'm not worried about that being inaccurate. The PoW mining I don't think we're going to get right, so I've just removed it. I think this is OK, as I think users will be more interested in Decred-specific terms, or general terms with a Decred twist (there are quite a few). If you Google PoW, plenty of good alternate definitions will give you an idea. Pretty sure we're doing the same as Bitcoin, just with a different hashing algorithm. I would suggest merging this branch, then I can open a PR to dcrdocs. If there are any terms we're still not sure on, we can flag them for further review in the PR comments.

RichardRed0x commented:
Cool, sounds good.
